### PR TITLE
policy: Add Port Range Support for Policies

### LINKED ISSUE #16622: Port ranges in network policies
The new `endPort` field for K8s' NetworkPolicies is in [alpha in v1.21](https://kubernetes.io/docs/concepts/services-networking/network-policies/#targeting-a-range-of-ports) and likely to become stable soon. It allows policies to target a range of ports instead of a single port.

Supporting this in Cilium will require changes in the datapath.

Tasks:
* [x] Add port range management mechanism
  * https://github.com/cilium/cilium/pull/20633
* [x] Add support for parsing/embedding port ranges from the API into the policy engine
* [x] Extend datapath for port range support
* [x] Implement policy engine -> datapath map generation logic

### PR DESCRIPTION
**Pull request by nathanjsweet**
See commits.

Fixes: #16622

```release-note
policy: Add support for port ranges in network policies.
```

### CODE REVIEW

**Review by chancez — APPROVED** [review]
Updates to Hubble parser tests LGTM.

File: pkg/envoy/xds_server.go [hunk 1]
```diff
@@ -1677,8 +1677,11 @@ func getDirectionNetworkPolicy(ep endpoint.EndpointUpdater, l4Policy policy.L4Po
 			continue
 		}
 
+		// NPDS supports port ranges, but range overlaps in the policy are NOT allowed.
+		// Envoy will reject a policy if any overlaps exist.
```
Comment (mhofstetter) [inline pkg/envoy/xds_server.go:1681]: it seems that Cilium Proxy (Envoy) now has support for overlapping port ranges: https://github.com/cilium/proxy/pull/373 ?

-> maybe adapt the comment accordingly?

**Review by mhofstetter — APPROVED** [review]
nice!

`cli` and `envoy` looks good to me :+1: 

PS: The relevant `cilium/proxy` PRs that added support for port ranges are https://github.com/cilium/proxy/pull/99 & https://github.com/cilium/proxy/pull/373

**Comment by joestringer** [comment]
:wave: Just looking from a high level meta approach, the PR is pretty big, coming into around 3.5K lines changed overall. Would it make sense to break this PR down into smaller chunks? As a reviewer I usually prefer that PRs are under 1K lines just to be able to keep things straight in my head, but I also undersand that several of the building blocks here only make sense in context of the broader context of the whole PR here. At a glance at least the bitlpm refactoring could be proposed+merged separately. That would hopefully have no functional impact and would trim ~500 lines or so from the rest to be reviewed in this PR.

**Review by learnitall — CHANGES_REQUESTED** [review]
I think that the changes to the PortProtocol API object need to be reflected in `./Documentation/security/policy/language.rst`. Can you take a look?

**Comment by nathanjsweet** [comment]
Going off of @joestringer recommendation, I will break this PR up. Here is the first part:
https://github.com/cilium/cilium/pull/32430

**Comment by nathanjsweet** [comment]
I'm turning this into a draft for the time being. This does not mean this PR is stalled; it is simply broken up. Part 1 is open for review. I'll open up part 2 as soon as part 1 is merged.

File: pkg/policy/portrange_test.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+			expected: []MaskedPort{
+				{port: 0xffff, mask: 0xffff}, // 65535
+			},
+		},
+		// These remaining tests are testing invalid cases where start >= end.
+		// For now, these cases return the start port with a full mask,
+		// indicating that only the start port should be part of the range.
+		// This is technically correct for the case of end port being "0"
+		// and the value of the port, but is ambiguous otherwise.
+		{
```
Comment (tamilmani1989) [inline pkg/policy/portrange_test.go:308]: What's the behavior if endport > 65535 but start is valid. can we also add a test for port range something like 50000-70000?

Comment (nathanjsweet) [reply pkg/policy/portrange_test.go:308]: The underlying type is a uint16 so it is not possible for the value to be greater than 65535. The 2nd part of this PR ([the first part is merged](https://github.com/cilium/cilium/pull/32430)) will validate that the API cannot send an EndPort larger than that to the policy engine.

**Pull request closed without merging**